### PR TITLE
Updates the update checker to run asynchronously.

### PR DESCRIPTION
**Pull request by binaek**

### CODE REVIEW

File: pkg/task/display.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+		// nothing to save
+		return nil
+	}
+
+	notifs := &Notifications{
+		StructVersion:           NotificationsStructVersion,
+		CLINotificationsLines:   cliNotificationsLines,
+		PluginNotificationLines: pluginNotificationLines,
+	}
+	if files.FileExists(filepaths.NotificationsFilePath()) {
```
Comment (binaek) [inline pkg/task/display.go:39]: Not sure if this is necessary!

File: cmd/root.go [hunk 2]
```diff
@@ -52,7 +62,13 @@ var rootCmd = &cobra.Command{
...
 		initGlobalConfig()
 
-		task.RunTasks()
+		t, err := task.RunTasks(cmd.Context(), cmd, args)
+		if err != nil {
+			log.Printf("[TRACE] ran into an error running daily tasks: %v", err)
+			log.Printf("[TRACE] stack: %s", string(debug.Stack()))
+		}
+		// save the channel so that we can wait on it when exiting
+		waitForTasks = t
```
Comment (binaek) [inline cmd/root.go:71]: update name to `tasksDoneChannel`

File: cmd/root.go [hunk 3]
```diff
@@ -52,7 +62,13 @@ var rootCmd = &cobra.Command{
 
 		initGlobalConfig()
 
-		task.RunTasks()
+		t, err := task.RunTasks(cmd.Context(), cmd, args)
```
Comment (binaek) [inline cmd/root.go:65]:
```
var xxx chan
var err error
```

File: cmd/root.go [hunk 4]
```diff
@@ -33,11 +33,28 @@ import (
...
 // rootCmd represents the base command when called without any subcommands
 var rootCmd = &cobra.Command{
 	Use:     "steampipe [--version] [--help] COMMAND [args]",
 	Version: version.SteampipeVersion.String(),
+	PersistentPostRun: func(cmd *cobra.Command, args []string) {
+		utils.LogTime("cmd.PersistentPostRun start")
+		defer utils.LogTime("cmd.PersistentPostRun end")
+
+		if waitForTasksChannel != nil {
+			defer tasksCancel()
```
Comment (kaidaguerre) [inline cmd/root.go:48]: do we need to cancel if we get a result on waitForTasksChannel?

File: pkg/task/display.go (hunk 1 above)
Comment (kaidaguerre) [reply pkg/task/display.go:40]: well - can you check whether it is and remove it not

File: pkg/task/display.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	f, err := os.Open(filepaths.NotificationsFilePath())
+	if err != nil {
+		return nil, err
+	}
+	notifications := &Notifications{}
+	decoder := json.NewDecoder(f)
+	if err := decoder.Decode(notifications); err != nil {
+		return nil, err
+	}
+	return notifications, error_helpers.CombineErrors(f.Close(), os.Remove(filepaths.NotificationsFilePath()))
```
Comment (kaidaguerre) [inline pkg/task/display.go:69]: split into multiple lines

Comment (kaidaguerre) [reply pkg/task/display.go:69]:
```
err := error_helpers.CombineErrors(f.Close(), os.Remove(filepaths.NotificationsFilePath()))
if err != nil{
    return nil, err
}
return notifications, nil
```

File: pkg/task/display.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// displayNotifications checks if there are any pending notifications to display
+// and if so, displays them
+// does nothing if the given command is a command where notifications are not displayed
+func (r *Runner) displayNotifications(cmd *cobra.Command, cmdArgs []string) error {
+	utils.LogTime("Runner.displayNotifications start")
+	defer utils.LogTime("Runner.displayNotifications end")
+
+	if isNoNotificationCmd(cmd, cmdArgs) {
```
Comment (kaidaguerre) [inline pkg/task/display.go:79]: rename to `showNotificationsForCommand` and reverse logic

File: pkg/task/display.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		// do not do anything
+		// just return
+		return nil
```
Comment (kaidaguerre) [inline pkg/task/display.go:82]: 2 comment lines is excessive here

File: pkg/task/runner.go [hunk 8]
```diff
@@ -111,7 +132,7 @@ func (r *Runner) shouldRun() bool {
 	return durationElapsedSinceLastCheck > minimumDurationBetweenChecks
 }
 
-func isIgnoredCmd(cmd *cobra.Command, cmdArgs []string) bool {
+func isNoNotificationCmd(cmd *cobra.Command, cmdArgs []string) bool {
```
Comment (kaidaguerre) [inline pkg/task/runner.go:135]: rename and add comment

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: pkg/task/display.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/turbot/steampipe/pkg/utils"
+)
+
+const (
+	NotificationsStructVersion = 20221121
+)
+
+type Notifications struct {
+	StructVersion           uint32   `json:"struct_version"`
+	CLINotificationsLines   []string `json:"cli_notifications"`
```
Comment (kaidaguerre) [inline pkg/task/display.go:23]: `CLINotificationLines` -> `CliNotifications` 
`PluginNotificationLines` -> `PluginNotifications`

File: pkg/task/display.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+		// nothing to display
+		return nil
+	}
+
+	notifications, err := r.getNotifications()
+	if err != nil {
+		return err
+	}
+
+	notificationLines := append(notifications.CLINotificationsLines, notifications.PluginNotificationLines...)
```
Comment (kaidaguerre) [inline pkg/task/display.go:98]: can we just update `r.getNotifications` to return a list of all notification lines, i.e. move the append code inside the function

File: pkg/task/runner.go [hunk 11]
```diff
@@ -20,84 +22,108 @@ type Runner struct {
...
+// returns a channel which is closed once all tasks are finished or the provided context is cancelled
+func RunTasks(ctx context.Context, cmd *cobra.Command, args []string) chan struct{} {
 	utils.LogTime("task.RunTasks start")
 	defer utils.LogTime("task.RunTasks end")
 
-	NewRunner().Run()
+	doneChannel := make(chan struct{}, 1)
+	runner := newRunner()
+
+	if err := runner.displayNotifications(cmd, args); err != nil {
```
Comment (kaidaguerre) [inline pkg/task/runner.go:34]: comment - `if there are any notifications from the previous run - display them`

File: pkg/task/runner.go [hunk 12]
```diff
@@ -20,84 +22,108 @@ type Runner struct {
...
+		log.Println("[TRACE] faced error displaying notifications:", err)
+	}
+
+	// asynchronously start the update checker
```
Comment (kaidaguerre) [inline pkg/task/runner.go:38]: // asynchronously run the task runner

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: pkg/task/display.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+	if err := decoder.Decode(notifications); err != nil {
+		return nil, err
+	}
+	if err := error_helpers.CombineErrors(f.Close(), os.Remove(filepaths.NotificationsFilePath())); err != nil {
+		// if Go couldn't close the file handle, no matter - this was just good practise
+		// if Go couldn't remove the notification file, it'll get truncated next time we try to write to it
+		// worst case is that the notification gets shown more than once
+		log.Println("[TRACE] could not close/delete notification file", err)
+	}
+	return append(notifications.CliNotifications, notifications.PluginNotification...), nil
```
Comment (kaidaguerre) [inline pkg/task/display.go:74]: why not add a GetAll function to Notifications

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

**Pull request merged**